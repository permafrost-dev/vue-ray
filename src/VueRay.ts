import { determineComponentNameDuringEvent } from '@/lib/helpers';
import { RayMixin } from '@/RayMixin';
import { format as prettyPrint } from '@permafrost-dev/pretty-format';
import { Ray } from 'node-ray/dist/web';
import { getCurrentInstance, type Ref, watch } from 'vue';

export class VueRay extends Ray {
    public component: any = { $data: {}, $refs: {} };
    public $watch: CallableFunction = () => {};

    public data(): void {
        if (this.component) {
            let data = Object.assign({}, this.component.data);

            // remove Vue Observers to avoid recursive issues
            delete data['trackingRays'];
            delete data['trackingStops'];

            data = Object.freeze(data);

            this.table([data], 'Data');
        }
    }

    public props(): void {
        if (this.component) {
            this.table([this.component.props]);
        }
    }

    public element(refName: string): VueRay {
        if (this.component?.refs[refName]) {
            return this.html((this.component?.refs[refName] as any).outerHTML);
        }

        return this;
    }

    public track(name: string): void {
        if (this.component && typeof this.component.trackingRays[name] === 'undefined') {
            this.component.trackingRays[name] = RayMixin.methods.$ray(this.component)(this.component.data['name']);

            const onTrackedUpdate = (value: any, previousValue: any) => {
                this.component.trackingRays[name] = this.component.trackingRays[name].sendCustom(
                    [
                        `<code style="font-weight: bold;">tracking: </code><code style="color: #16a34a;">${determineComponentNameDuringEvent(this.component.type)}.data.${name}</code>`,
                        `<hr width="33%" align="left" /><code style="color: #94a3b8;">new value: </code><code class="text-gray-500 bold font-bold">${prettyPrint(value)}</div>`,
                        `                               <code style="color: #94a3b8;">old value: </code><code class="text-gray-700">${prettyPrint(previousValue)}</code>`,
                        `<hr width="33%" align="left" /><code style="color: #94a3b8;">changed at: </code><code style="color: #60a5fa;">${new Date().toISOString()}</div>`,
                    ].join('<br>'),
                    'Tracking Data',
                );
            };

            this.component.trackingStops[name] = this.$watch(() => this.component.data[name], onTrackedUpdate, { deep: true });
        }
    }

    public untrack(name: string): void {
        if (this.component && typeof this.component.trackingStops[name] !== 'undefined') {
            this.component.trackingStops[name]();
            delete this.component.trackingStops[name];
        }

        if (this.component && typeof this.component.trackingRays[name] !== 'undefined') {
            delete this.component.trackingRays[name];
        }
    }

    public unwatch(name: string): void {
        if (this.component && typeof this.component.trackingStops[name] !== 'undefined') {
            this.component.trackingStops[name]();
            delete this.component.trackingStops[name];
        }

        if (this.component && typeof this.component.trackingRays[name] !== 'undefined') {
            delete this.component.trackingRays[name];
        }
    }

    public watch(name: string, ref: Ref<any>): void {
        if (this.component && typeof this.component.trackingRays[name] === 'undefined') {
            this.component.trackingRays[name] = RayMixin.methods.$ray(this.component)(ref.value);

            const onTrackedUpdate = (value: any, previousValue: any) => {
                this.component.trackingRays[name] = this.component.trackingRays[name].sendCustom(
                    [
                        `<code style="font-weight: bold;">watching: </code><code style="color: #16a34a;">${determineComponentNameDuringEvent(this.component.type)} âžž Ref&lt;${name}&gt;</code>`,
                        `<hr width="33%" align="left" /><code style="color: #94a3b8;">new value: </code><code class="text-gray-500 bold font-bold">${prettyPrint(value)}</div>`,
                        `                               <code style="color: #94a3b8;">old value: </code><code class="text-gray-700">${prettyPrint(previousValue)}</code>`,
                        `<hr width="33%" align="left" /><code style="color: #94a3b8;">changed at: </code><code style="color: #60a5fa" class="text-gray-500 bold font-bold">${new Date().toISOString()}</div>`,
                    ].join('<br>'),
                    'Watching Ref',
                );
            };

            this.component.trackingStops[name] = this.$watch(ref, onTrackedUpdate, { deep: true });
        }
    }

    public static useDefaultSettings(settings: any) {
        return super.useDefaultSettings(settings);
    }

    public send(...args: any[]) {
        return super.send(...args);
    }
}

export const ray = (...args: any[]) => {
    const result = VueRay.create() as VueRay;
    result.component = getCurrentInstance();
    result.$watch = watch;

    if (!args.length) {
        return result;
    }

    result.send(...args);

    return result;
};

/**
 * Returns a ray() function that is wrapped with the component instance
 *
 * @param component
 * @returns {(...args: any[]) => VueRay}
 */
// eslint-disable-next-line no-unused-vars
export function rayWrapped(component: any): (...args: any[]) => VueRay {
    // biome-ignore lint/complexity/useArrowFunction: <explanation>
    return function (...args: any[]) {
        const result = VueRay.create() as VueRay;

        result.component = component;
        result.$watch = watch;

        if (!args.length) {
            return result;
        }

        result.send(...args);

        return result;
    };
}
