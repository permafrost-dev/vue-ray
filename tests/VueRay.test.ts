import { VueRay } from '@/VueRay';
import { beforeEach, describe, expect, it } from 'vitest';
import { FakeClient } from './TestClasses/FakeClient';

interface WatchParameters {
    cb1: CallableFunction;
    cb2: CallableFunction;
    options: Record<string, unknown>;
}

let ray: VueRay,
    client: FakeClient,
    lastWatchParams: WatchParameters | null = null;

beforeEach(() => {
    client = new FakeClient(3000, 'localhost');
    // @ts-ignore
    ray = VueRay.create(client, 'fakeUuid'); // ({ enable: true, port: 3000, host: 'localhost' }, client, 'fakeUuid');

    ray.component = {
        $data: { one: 1 },
        $props: { title: 'abc' },
        $refs: { test1: { innerHTML: '<em>123</em>' } },
        trackingRays: {},
        trackingStops: {},
        $ray(...args: any[]) {
            // @ts-ignore
            return ray.send(...args);
        },
    };

    ray.watch = (cb1: CallableFunction, cb2: CallableFunction, options: Record<string, unknown> = {}) => {
        lastWatchParams = {
            cb1,
            cb2,
            options,
        };

        return () => true;
    };

    ray.data();
});

describe('Base VueRay class:', () => {
    it('sends the component data', () => {
        ray.data();

        expect(client.sentRequests).toMatchSnapshot();
    });

    it('sends the component props', () => {
        ray.props();

        expect(client.sentRequests).toMatchSnapshot();
    });

    it('sends a refs innerHTML', () => {
        ray.component = { refs: { test1: { innerHTML: '<em>123</em>' } } };
        ray.element('test1');

        expect(client.sentRequests).toMatchSnapshot();
    });

    it('sets the show_component_lifecycles property', () => {
        VueRay.showComponentLifecycles(['one', 'two']);

        expect(VueRay.show_component_lifecycles).toStrictEqual(['one', 'two']);
    });

    // it('searches the show_component_lifecycles property', () => {
    //     VueRay.showComponentLifecycles(['one', 'two']);

    //     expect(VueRay.shouldDisplayEvent('one')).toBeTruthy();
    //     expect(VueRay.shouldDisplayEvent('three')).toBeFalsy();
    // });

    it('clears the show_component_lifecycles property', () => {
        VueRay.showComponentLifecycles(['one', 'two']);
        VueRay.stopShowingComponentLifecycles([]);

        expect(VueRay.show_component_lifecycles.length).toBe(0);
    });

    it('removes items from the show_component_lifecycles property', () => {
        VueRay.showComponentLifecycles(['one', 'two']);
        VueRay.stopShowingComponentLifecycles(['two']);

        expect(VueRay.show_component_lifecycles).toStrictEqual(['one']);
    });

    it('tracks data variables', () => {
        ray.track('test');

        lastWatchParams?.cb2();

        expect(client.sentRequests).toMatchSnapshot();
    });

    it('stops tracking data variables', () => {
        ray.track('test');

        expect(ray.component.trackingRays['test']).toBeDefined();
        expect(ray.component.trackingStops['test']).toBeDefined();

        ray.untrack('test');

        expect(ray.component.trackingRays['test']).toBeUndefined();
        expect(ray.component.trackingStops['test']).toBeUndefined();

        expect(client.sentRequests).toMatchSnapshot();
    });
});
