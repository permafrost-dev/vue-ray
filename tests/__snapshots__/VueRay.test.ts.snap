// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Base VueRay class: > sends a refs innerHTML 1`] = `
[
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "label": "Data",
          "values": [
            "<code style="font-size: 0.8rem!important;" class=""><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<code style="font-size: 0.8rem!important;" class="text-gray-600"><span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"one"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span>1<span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br><span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span></code></code>",
          ],
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "table",
      },
    ],
    "uuid": "fakeUuid",
  },
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "content": "<em>123</em>",
          "label": "HTML",
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "custom",
      },
    ],
    "uuid": "fakeUuid",
  },
]
`;

exports[`Base VueRay class: > sends the component data 1`] = `
[
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "label": "Data",
          "values": [
            "<code style="font-size: 0.8rem!important;" class=""><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<code style="font-size: 0.8rem!important;" class="text-gray-600"><span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"one"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span>1<span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br><span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span></code></code>",
          ],
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "table",
      },
    ],
    "uuid": "fakeUuid",
  },
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "label": "Data",
          "values": [
            "<code style="font-size: 0.8rem!important;" class=""><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<code style="font-size: 0.8rem!important;" class="text-gray-600"><span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"one"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span>1<span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br><span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span></code></code>",
          ],
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "table",
      },
    ],
    "uuid": "fakeUuid",
  },
]
`;

exports[`Base VueRay class: > sends the component props 1`] = `
[
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "label": "Data",
          "values": [
            "<code style="font-size: 0.8rem!important;" class=""><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<code style="font-size: 0.8rem!important;" class="text-gray-600"><span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"one"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span>1<span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br><span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span></code></code>",
          ],
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "table",
      },
    ],
    "uuid": "fakeUuid",
  },
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "label": "Table",
          "values": [
            "<code style="font-size: 0.8rem!important;" class=""><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<code style="font-size: 0.8rem!important;" class="text-gray-600"><span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"title"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span><code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"abc"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br><span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span></code></code>",
          ],
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "table",
      },
    ],
    "uuid": "fakeUuid",
  },
]
`;

exports[`Base VueRay class: > stops tracking data variables 1`] = `
[
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "label": "Data",
          "values": [
            "<code style="font-size: 0.8rem!important;" class=""><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<code style="font-size: 0.8rem!important;" class="text-gray-600"><span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"one"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span>1<span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br><span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span></code></code>",
          ],
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "table",
      },
    ],
    "uuid": "fakeUuid",
  },
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "values": [
            "test",
          ],
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "log",
      },
    ],
    "uuid": "fakeUuid",
  },
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "content": "<code style="font-size: 0.8rem!important;" class=""><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<code style="font-size: 0.8rem!important;" class="text-gray-600"><span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"info"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"tracking"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span><span style="font-size: 0.8rem!important;" class="bold text-indigo-600">false</span><span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"variable"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span><code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"test"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span><span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"trackedData"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"value"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span><span style="font-size: 0.8rem!important;" class="bold text-indigo-600">undefined</span><span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span><span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br><span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span></code></code>",
          "label": "HTML",
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "custom",
      },
    ],
    "uuid": "fakeUuid",
  },
]
`;

exports[`Base VueRay class: > tracks data variables 1`] = `
[
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "label": "Data",
          "values": [
            "<code style="font-size: 0.8rem!important;" class=""><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<code style="font-size: 0.8rem!important;" class="text-gray-600"><span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"one"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span>1<span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br><span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span></code></code>",
          ],
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "table",
      },
    ],
    "uuid": "fakeUuid",
  },
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "values": [
            "test",
          ],
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "log",
      },
    ],
    "uuid": "fakeUuid",
  },
  {
    "meta": [],
    "payloads": [
      {
        "content": {
          "content": "<code style="font-size: 0.8rem!important;" class=""><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<code style="font-size: 0.8rem!important;" class="text-gray-600"><span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"info"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"tracking"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span><span style="font-size: 0.8rem!important;" class="bold text-indigo-600">true</span><span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"variable"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span><code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"test"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span><span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"trackingData"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span><span style="font-size: 0.8rem!important;" class="bold text-yellow-600">Object</span>&nbsp;<span style="font-size: 0.8rem!important;" class="bold text-orange-400">{</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"previous"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span><span style="font-size: 0.8rem!important;" class="bold text-indigo-600">undefined</span><span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1&nbsp;&nbsp;&nbsp;&nbsp;1<code style="font-size: 0.8rem!important;" class="bold text-green-600 p-0">"value"</code><span style="font-size: 0.8rem!important;" class="bold text-orange-400">:&nbsp;</span><span style="font-size: 0.8rem!important;" class="bold text-indigo-600">undefined</span><span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br>&nbsp;&nbsp;&nbsp;&nbsp;1<span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span><span style="font-size: 0.8rem!important;" class="bold text-orange-400">,</span><br><span style="font-size: 0.8rem!important;" class="bold text-orange-400">}</span></code></code>",
          "label": "HTML",
        },
        "origin": {
          "file": "web.cjs.js",
          "function_name": "xxxx",
          "hostname": "remote",
          "line_number": 999,
        },
        "type": "custom",
      },
    ],
    "uuid": "fakeUuid",
  },
]
`;
